segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,436 @@
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        Path path = Paths.get(config.getRoot(), streamSegmentName);
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        } else if (Files.isWritable(path)) {
+            LoggerHelpers.traceLeave(log, "openRead", traceId);
+            return FileSystemSegmentHandle.writeHandle(streamSegmentName);
+        } else {
+            LoggerHelpers.traceLeave(log, "openWrite", traceId);
+            return syncOpenRead(streamSegmentName);
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) {
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle.getSegmentName(), offset, bufferOffset, length);
+
+        Path path = Paths.get(config.getRoot(), handle.getSegmentName());
+
+        long fileSize = Files.size(path);
+        if (fileSize < offset) {
+            throw new IllegalArgumentException(String.format("Reading at offset (%d) which is beyond the " +
+                    "current size of segment (%d).", offset, fileSize));
+        }
+
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+            int bytesRead = 0;
+
+            do {
+                ByteBuffer readBuffer = ByteBuffer.wrap(buffer, bufferOffset, length);
+                bytesRead = FileSystemRetryHelper.retry(
+                        () -> channel.read(readBuffer, offset),
+                        (read) -> read <= 0,
+                        () -> new IllegalStateException("Retries exceeded while Reading"),
+                        NUM_RETRIES);
+
+                bufferOffset += bytesRead;
+                length -= bytesRead;
+            } while (length != 0);
+            LoggerHelpers.traceLeave(log, "read", traceId, bytesRead);
+            return bytesRead;
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private SegmentProperties syncGetStreamSegmentInfo(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+        PosixFileAttributes attrs = Files.readAttributes(Paths.get(config.getRoot(), streamSegmentName),
+                PosixFileAttributes.class);
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName, attrs.size(),
+                !(attrs.permissions().contains(OWNER_WRITE)), false,
+                new ImmutableDate(attrs.creationTime().toMillis()));
+
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        return Files.exists(Paths.get(config.getRoot(), streamSegmentName));
+    }
+
+    @SneakyThrows
```

Comment:
Why are we sneaky-throwing generally rather than specifying the exception? I'm not the best person to ask about it, but I know for sure that it needs to be used judiciously.